network/local.go

```diff
@@ -0,0 +1,487 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// NewLocalRouter returns a fresh router which uses only local queues. It uses
+// the default local manager.
+// If you need multiple independent local-queues, use NewLocalRouterWithManager.
+// In case of an error it is returned together with a nil-Router.
+func NewLocalRouter(sid *ServerIdentity) (*Router, error) {
+	return NewLocalRouterWithManager(defaultLocalManager, sid)
+}
+
+// NewLocalRouterWithManager is the same as NewLocalRouter but takes a specific
+// LocalManager. This is useful to run parallel different local overlays.
+// In case of an error it is returned together with a nil-Router.
+func NewLocalRouterWithManager(lm *LocalManager, sid *ServerIdentity) (*Router, error) {
+	h, err := NewLocalHostWithManager(lm, sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// LocalManager keeps a reference to all opened local connections.
+// It also keeps track of who is "listening", so it's possible to mimic
+// Conn & Listener.
+type LocalManager struct {
+	// queues maps a remote endpoint to its packet queue. It's the main
+	// structure used to communicate.
+	queues map[endpoint]*connQueue
+	sync.Mutex
+	// The listening-functions used when a new connection-request arrives.
+	listening map[Address]func(Conn)
+
+	// connection-counter for giving unique IDs to each connection.
+	counter uint64
+}
+
+// NewLocalManager returns a fresh new manager that can be used by LocalConn,
+// LocalListener & LocalHost.
+func NewLocalManager() *LocalManager {
+	return &LocalManager{
+		queues:    make(map[endpoint]*connQueue),
+		listening: make(map[Address]func(Conn)),
+	}
+}
+
+// defaultLocalManager can be used if you need only one LocalManager.
+var defaultLocalManager = NewLocalManager()
+
+// endpoint represents one endpoint of a connection.
+type endpoint struct {
+	// addr is the Address of this endpoint.
+	addr Address
+	// uid is a unique identifier of the remote endpoint
+	// it's unique  for each direction:
+	// 127.0.0.1:2000 -> 127.0.0.1:2000 => 14
+	// 127.0.0.1:2000 <- 127.0.0.1:2000 => 15
+	uid uint64
+}
+
+// LocalReset resets the map of connections + listeners for the defaultLocalManager.
+func LocalReset() {
+	defaultLocalManager = NewLocalManager()
+
+}
+
+// isListening returns true if the remote address is listening for connections.
+func (lm *LocalManager) isListening(remote Address) bool {
+	lm.Lock()
+	defer lm.Unlock()
+	_, ok := lm.listening[remote]
+	return ok
+}
+
+// setListening marks the address as being able to accept incoming connections.
+// For each incoming connection, fn will be called in a go routine.
+func (lm *LocalManager) setListening(addr Address, fn func(Conn)) {
+	lm.Lock()
+	defer lm.Unlock()
+	lm.listening[addr] = fn
```

Comment:
I like the defer thing, since you can directly check if  the mutex is "correctly" released: `it locks and make sure at the end it is unlocked`. Ok here this is one line above / below, but still. And being consistent on one way of doing one thing is also good I think.